### ISSUE TITLE
Interpreting stack/cabal projects

### ISSUE THREAD

**Issue by m0rphism** [mvdan/hint#25]
I think it would be great to have functions to initialize a hint session from a `.cabal` or `stack.yaml` file,
the same way as `cabal repl` or `stack repl` initializes `ghci`.

The functions could be for example:

    runInterpreterWithStackFile, runInterpreterWithCabalFile
      :: (MonadIO m, MonadMask m)
      => FilePath
      -> InterpreterT m a
      -> m (Either InterpreterError a)

I've tried to implement that functionality myself by running `stack repl --with-ghc echo` to get the command line arguments normally passed to `ghci`, and forwarded them to `unsafeRunInterpreterWithArgs`, after removing the `--interactive` argument, which seems to be rejected by hint. With this approach I was able to interpret an example stack project, which was depending on another external package. The approach seems not very robust, though, and I'm not convinced how well it works in general. ;-)

Thanks for working on `hint`, by the way! :+1: 
I'm currently trying to use it in my text editor to load a Haskell configuration project, which can depend on compiled editor plugins. :-)

**Comment by gelisam** [comment]
+1, that would be useful for my project as well

**Comment by mvdan** [comment]
If you send a PR, I'll be happy to review :)

**Comment by m0rphism** [comment]
Great that you guys are positive towards the idea! :-)

I've implemented a first prototype for `runInterpreterWithStackFile` here: 8381efdf969349c1e9e38d43d806b29989a4d5c6

However, I'm not very happy with the approach of parsing the output of `stack --with-ghc=echo`.
This causes an ugly dependency on how the `stack` executable prints to stdout.

I've reexported the functions in the `Language.Haskell.Interpreter.Unsafe` module, as `stack` forwards the `ghc-options` from the `.cabal` file to `ghci`, so I guess the same applies as for `unsafeRunInterpreterWithArgs`.

I wonder if it would be reasonable to use `stack` as a library instead. This may offer safety and speed, but on the other hand would add a big dependency. The latter could be mediated by having a `hint-stack` and `hint-cabal` package, so `hint`'s dependencies can remain small.

I just had a quick look at the `stack` code base, but wasn't able to find a direct way for retrieving the `ghci` arguments. However, it might be not that difficult to reproduce the corresponding part from the `stack` executable. [Here](https://github.com/commercialhaskell/stack/blob/99dad1e45da0599e79703d068dddef0bc755de51/src/Stack/Ghci.hs#L354) is the part in `stack` which creates the `ghci` arguments.

**Comment by mvdan** [comment]
I agree that adding the dependency to hint itself would not be a good idea.

Hint itself doesn't depend on `ghc` being readily executable, so it would be unfortunate if now we depended on other tools like `cabal` or `stack` to be. Perhaps separate packages like you suggest would be the best way forward.

I have no idea about the `repl` command or much knowledge about cabal/stack in general, so my design input is fairly limited.

**Comment by m0rphism** [comment]
> I agree that adding the dependency to hint itself would not be a good idea.
> 
> Hint itself doesn't depend on ghc being readily executable, so it would be unfortunate if now we depended on other tools like cabal or stack to be. Perhaps separate packages like you suggest would be the best way forward.

That makes sense.

What do you think about the approach used in the current prototype with spawning processes?
The only new static dependency would be `process`, but the `stack`-specific `hint` functions would require a `stack` executable on the runtime `PATH` to succeed. The main belly ache I get from this approach, is the whole *stringyness* and cross-platform support, but maybe those are wrong instincts in this case ;-)

> I have no idea about the repl command or much knowledge about cabal/stack in general, so my design input is fairly limited.

I'm pretty much in the same situation. :-)

My hopes are that the problem can be solved by reusing much of the infrastructure already provided for `ghci`. Maybe we could get some input on the stack mailing list.

By the way, what exactly makes `unsafeRunInterpreterWithArgs` unsafe?
Are there arguments (besides `--interactive`) that `ghci` handles but `hint` doesn't?

**Comment by mvdan** [comment]
> But I also completely understand, if you want to avoid such runtime dependencies in an otherwise pure package for consistency.

I assume if they are separate packages, any other packages could depend on and only build what they need? For example, current users of `hint` could ignore `hint-cabal` and `hint-stack`. If that is possible, I'd definitely prefer using cabal/stack as libraries at compile time.

If it turns out to be impossible or ridiculously complex, then perhaps hacky command execution at runtime is a good option.

> By the way, what exactly makes unsafeRunInterpreterWithArgs unsafe?

See https://github.com/mvdan/hint/pull/4. Basically, it's possible to make mistakes (e.g. typos in the type names) and the compiler won't catch them. Otherwise, the compiler will actually resolve the types.

As seen in the PR, unsafe is more powerful because, for example, it lets you come up with the types at runtime, not at compile time. Think of it as some sort of code generation. That's not possible at compile time, unless you have boilerplate for every single possible type.

**Comment by gelisam** [comment]
These days, a more modern way to accomplish this would be to use https://github.com/mpickering/hie-bios to convert a stack or cabal project into the set of flags we need to pass to ghc.

**Comment by gelisam** [comment]
(I didn't remove the "help wanted" tag because help is no longer wanted, but because that tag is redundant; I welcome help on all the issues, not just those which have the "help wanted" tag!)
